### PR TITLE
Add diagnostics to "while loop" and "for loop" that note that it is always determined that it might iterate zero times.

### LINKED ISSUE #116572: Logically identical code. One compiles, one does not
I tried to compile this code [playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=3a7b73546d419d316ad792d0c6362152):

```
use std::time::{Duration, Instant};
use std::thread;

fn time_bounded_retry<T, E>(
    max_time: Duration,
    sleep_for: Duration,
    f: impl Fn() -> Result<T, E>,
) -> Result<T, E> {
    let start = Instant::now();
    while let result = f() {
        match result {
            Ok(_) => return result,
            Err(_) => {
                if start.elapsed() > max_time {
                    return result;
                }

                thread::sleep(sleep_for);
            }
        }
    }
    // unreachable!()
}
```

However, it produces an error. If you uncomment the `unreachable!()` will work. Here's the error message:

```
    |
191 |     while let result = f() {
    |     ^^^^^^^^^^^^^^^^^^^^^^ this might have zero elements to iterate on
192 |         match result {
193 |             Ok(_) => return result,
    |                      ------------- if the loop doesn't execute, this value would never get returned
...
196 |                     return result;
    |                     ------------- if the loop doesn't execute, this value would never get returned
    = help: return a value for the case when the loop has zero elements to iterate on, or consider changing the return type to account for that possibility
help: try adding an expression at the end of the block
```

This code is functionally identical to the `while let` code and can compile without an `unreachable!()`:

```
fn time_bounded_retry<T, E>(
    max_time: Duration,
    sleep_for: Duration,
    f: impl Fn() -> Result<T, E>,
) -> Result<T, E> {
    let start = Instant::now();
    loop {
        let result = f();
        match result {
            Ok(_) => return result,
            Err(_) => {
                if start.elapsed() > max_time {
                    return result;
                }

                thread::sleep(sleep_for);
            }
        }
    }
}
```

The code in the while loop is irrefutable, so it is functionally the same as the `loop` case. Even if it's not a good practice to use an irrefutable statement in a while loop (and clippy will catch this) I think it should still compile.

My main reason for wanting it to compile is that while iterating, I wrote the `while let` version first, and because I could not get it to compile without the `unreachable!()` statement mistakenly believed that the compiler would not allow something like the second result. As a result of that mistaken impression ended up with this code:

```
fn time_bounded_retry<T, E>(
    max_time: Duration,
    sleep_for: Duration,
    f: impl Fn() -> Result<T, E>,
) -> Result<T, E> {
    let start = Instant::now();
    let mut result;
    loop {
        result = f();
        if result.is_ok() || start.elapsed() > max_time {
            break;
        }
        thread::sleep(sleep_for);
    }
    result
}
```

Which is not ideal (needs a temp mutable variable). A co-worker without the above learned helplessness `unreachable!()` experience wrote this code below, which I believe is a much better way to implement the logic as it does not require a temp mutable variable:

```
fn time_bounded_retry<T, E>(
    max_time: Duration,
    sleep_for: Duration,
    f: impl Fn() -> Result<T, E>,
) -> Result<T, E> {
    let start = Instant::now();
    loop {
        result = f();
        if result.is_ok() || start.elapsed() > max_time {
            return result;
        }
        thread::sleep(sleep_for);
    }
}
```

So effectively, I'm advocating for `while let var = //...` to be treated the same as `loop { let var = //...`, not because I think it's a good idea, but because having one error and not the other gives false feedback to the coder.

### Meta

`rustc --version --verbose`:
```
$ rustc --version --verbose
rustc 1.73.0 (cc66ad468 2023-10-03)
binary: rustc
commit-hash: cc66ad468955717ab92600c770da8c1601a4ff33
commit-date: 2023-10-03
host: x86_64-apple-darwin
release: 1.73.0
LLVM version: 17.0.2
```

### PR DESCRIPTION
**Pull request by Kohei316**
close #116572
When a user writes a loop expression as a tail expression, which has a loop header always executed, and which contains return expressions like this,
```rust
fn foo() -> bool {
    while let x = 0 {
        return true
    }
}
```
this,
```rust
fn foo() -> bool {
    while true {
        return true
    }
}
```
or this,
```rust
fn foo() -> bool {
    for i in 0..10 {
        return true
    }
}
```
they may be misundrestanding that it evaluate to not a unit type but un argument of the return expression.
This PR adds diagnositics which help clearing such misundrestanding.

### CODE REVIEW

File: compiler/rustc_hir_typeck/src/coercion.rs [hunk 1]
```diff
@@ -1715,6 +1717,33 @@ impl<'tcx, 'exprs, E: AsCoercionSite> CoerceMany<'tcx, 'exprs, E> {
...
+                            let loop_header = if loop_src == hir::LoopSource::While {
+                                "loop condition expression or pattern"
+                            } else {
+                                "iterator expression"
+                            };
+                            err.span_note(
+                                    loop_span,
+                                    format!("It is determined that this might iterate zero times, regardless of the {loop_header}.")
+                                );
+                            err.span_help(loop_span, "If you are assuming that it will iterate more than once, consider using a `loop` expression instead.");
```
Comment (zachs18) [inline compiler/rustc_hir_typeck/src/coercion.rs:1745]:
```suggestion
                            err.span_help(loop_span, "If you are assuming that it will iterate at least once, consider using a `loop` expression instead.");
```

File: tests/ui/typeck/irrefutable-while-let-pattern.stderr [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+   |
+LL |     while let x = 0 {
+   |     ^^^^^^^^^^^^^^^
+help: consider returning a value here
+   |
+LL ~     }
+LL +     /* `bool` value */
+   |
+
+error: aborting due to 1 previous error
```
Comment (nnethercote) [inline tests/ui/typeck/irrefutable-while-let-pattern.stderr:29]: A number of issues here.
- Error messages don't start with an upper-case letter, and don't end with a '.'.
- There is a lot of repetition here. The `note` and the following `help` both print the loop condition. The two could be merged. I am also not sure if the condition needs to be printed, given that the loop is printed above anyway.
- "It is determined" is too wordy, something like "this loop might iterate zero times" would be better.

But there are bigger concerns, which I will discuss below.

**Comment by nnethercote** [comment]
There's a big problem here: the error messages mentioned in #116572 have since changed greatly. E.g. for the first example in that issue it now just says "`while` loops evaluate to unit type `()`" which I find sufficiently informative. In #116572 @Nadrieril wrote:

> A workable solution would be instead to change the error message: the code that emits "this might have zero elements to iterate on" could check if the pattern is a single binding and emit a better error message if that's the case.

But because the message has changed so much, there is now no mention of zero elements, so this PR isn't following that suggestion.

#122561 was also complaining about the old message, and #123125 was the PR that changed the message in order to fix #122561. So I *think* this PR is not actually needed, and #116572 can now be closed. But I will ask Esteban, who reviewed #123125, for a second opinion, because this is a complicated situation and is stretching my expertise on error messages to breaking point.

r? @estebank

**Comment by estebank** [comment]
I will be gone for 3 weeks. I will look at this again when I'm back. I will need more time to provide a good review than I have right now. Apologies for the delay on getting to this. I also think that we're in a state where the ticket can be closed, but I think there's some value with telling people to use `loop`+`break`/`return`. Thankfully this isn't time sensitive.

**Comment by wesleywiser** [comment]
Hi @estebank, just a friendly ping from the compiler team triage meeting in case this slipped off your todo list 🙂

**Comment by pnkfelix** [comment]
I'd like to note that from my playing around with the original playground: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=3a7b73546d419d316ad792d0c6362152 it seems to me like the problem here still isn't addressed.

Consider this (a simplfied version of #116572):

```rust
fn time_bounded_retry() -> i32 {
    while let result = 3 { return 4; }
}
```

compared to this:

```rust
fn time_bounded_retry() -> i32 {
    while let result = 3 { return 4; }
    loop { }
}
```

The first gives a message saying the while loop has type unit.

The second gives a *much better* diagnostic, pointing out that the while condition is irrefutable, and that the person should consider writing it instead as:

```rust
fn time_bounded_retry() -> i32 {
    loop { let result = 3; return 4; }
    loop { }
}
```

I think we should give a diagnostic in the first scenario that is more on par with the diagnostic given in the second scenario.

**Comment by joan38** [comment]
What happened here?
